### PR TITLE
Use regex-based version casting that accept 

### PR DESCRIPTION
**Pull request by ccl-core**
Remove warning when version doesn't conform to MAJOR.MINOR.PATCH

Ref. #https://github.com/mlcommons/croissant/issues/609

### CODE REVIEW

File: python/mlcroissant/mlcroissant/_src/core/issues.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ class Issues:
     We use sets to represent errors and warnings to avoid repeated strings.
     """
 
+    ignore_warnings: bool = False
```
[the remark this answers is not in the document]
Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/core/issues.py:37]: +1 to put it in `report`.

File: python/mlcroissant/mlcroissant/_src/core/issues_test.py [hunk 2]
```diff
@@ -42,3 +42,22 @@ def test_issues():
...
+    assert not issues.errors
+    assert not issues.warnings
+
+    metadata = Metadata(id="abc", name="abc")
+    file_object = FileObject(id="xyz", name="xyz")
+    file_object.parents = [metadata]
+    issues.add_error("foo", metadata)
+    issues.add_warning("bar", file_object)
+
+    print(issues.report())
```
Comment (marcenacp) [inline python/mlcroissant/mlcroissant/_src/core/issues_test.py:58]: Remove?

**Comment by ccl-core** [comment]
That's right, this would actually address the "However, adding a generic --ignore framework is probably a good first step, so I'm opening this issue." part of the issue, but I think a more long-term solution might be needed to silence only a subset of the warnings.

File: python/mlcroissant/mlcroissant/_src/core/versions.py [hunk 3]
```diff
@@ -19,10 +19,9 @@ def cast_version(version: Any) -> str | None:
     if isinstance(version, str):
         numbers = version.split(".")
         are_not_all_numbers = any(not number.isnumeric() for number in numbers)
-        if len(numbers) != 3 or are_not_all_numbers:
+        if are_not_all_numbers:
             raise WarningException(
-                f"Version doesn't follow MAJOR.MINOR.PATCH: {version}. For more"
-                " information refer to: https://semver.org/spec/v2.0.0.html"
+                f"Version contains non-numeric characters: {version}."
```
Comment (ccl-core) [inline python/mlcroissant/mlcroissant/_src/core/versions.py:24]: Keeping the warning for non-numeric versioning. Reasonable? Or shall I remove this as well?

**Comment by ccl-core** [comment]
@marcenacp updated the PR -- PTAL, thanks! :)

File: python/mlcroissant/mlcroissant/_src/core/versions.py (hunk 3 above)
Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/core/versions.py:24]: We could fully implement the [semantic versioning standard](https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions).

[`semver`](https://pypi.org/project/semver) is a python library that does exactly this. However it seems we only need to check the compatibility with the standard, so we could also just reuse their [regular expression](https://github.com/python-semver/python-semver/blob/c47774db88386f7db79e5ec9fe3672c07ed7c063/src/semver/version.py#L81-L104).

File: python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata_test.py [hunk 4]
```diff
@@ -101,17 +101,9 @@ def test_invalid_version(version, expected_error):
 @pytest.mark.parametrize(
     ["version", "expected_warning"],
     [
-        [
```
Comment (marcenacp) [inline python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata_test.py:104]: Still keep this?

Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata_test.py:104]: (Those should fail)

Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata_test.py:104]: Above, you could add that the versions `1` and `1.2` now succeed

**Review by marcenacp — APPROVED** [review]

File: python/mlcroissant/mlcroissant/_src/core/versions.py (hunk 3 above)
Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/core/versions.py:24]: Also, I think that it's worth mentioning the semantic versioning standard: https://semver.org/spec/v2.0.0.html, and that minor and patch can be omitted.

**Review by marcenacp — CHANGES_REQUESTED** [review]
See comments.

**Review by marcenacp — COMMENTED** [review]
Thanks!

**Review by marcenacp — APPROVED** [review]

**Pull request merged**